src/System.IO.FileSystem/tests/FileStream/ReadAsync.cs

```diff
@@ -0,0 +1,329 @@
+
+                // out of bounds checking happens first
+                Assert.Throws<ArgumentException>(null, () => 
+                    FSAssert.CompletesSynchronously(fs.ReadAsync(new byte[2], 1, 2)));
+
+                // count is checked prior
+                Assert.Throws<ArgumentOutOfRangeException>("count", () => 
+                    FSAssert.CompletesSynchronously(fs.ReadAsync(new byte[1], 0, -1)));
+
+                // offset is checked prior
+                Assert.Throws<ArgumentOutOfRangeException>("offset", () => 
+                    FSAssert.CompletesSynchronously(fs.ReadAsync(new byte[1], -1, -1)));
+
+                // buffer is checked first
+                Assert.Throws<ArgumentNullException>("buffer", () => 
+                    FSAssert.CompletesSynchronously(fs.ReadAsync(null, -1, -1)));
+            }
+        }
+
+        [Fact]
+        public void CancelledTokenFastPath()
+        {
+            CancellationTokenSource cts = new CancellationTokenSource();
+            cts.Cancel();
+            CancellationToken cancelledToken = cts.Token;
+
+            using (FileStream fs = new FileStream(GetTestFilePath(), FileMode.Create))
+            {
+                FSAssert.IsCancelled(fs.ReadAsync(new byte[1], 0, 1, cancelledToken), cancelledToken);
+            }
+
+            using (FileStream fs = new FileStream(GetTestFilePath(), FileMode.Create, FileAccess.Write))
+            {
+                // before write only check
+                FSAssert.IsCancelled(fs.ReadAsync(new byte[1], 0, 1, cancelledToken), cancelledToken);
+
+                fs.Dispose();
+                // before disposed check
+                FSAssert.IsCancelled(fs.ReadAsync(new byte[1], 0, 1, cancelledToken), cancelledToken);
+
+                // out of bounds checking happens first
+                Assert.Throws<ArgumentException>(null, () =>
+                    FSAssert.CompletesSynchronously(fs.ReadAsync(new byte[2], 1, 2, cancelledToken)));
+
+                // count is checked prior
+                Assert.Throws<ArgumentOutOfRangeException>("count", () =>
+                    FSAssert.CompletesSynchronously(fs.ReadAsync(new byte[1], 0, -1, cancelledToken)));
+
+                // offset is checked prior
+                Assert.Throws<ArgumentOutOfRangeException>("offset", () =>
+                    FSAssert.CompletesSynchronously(fs.ReadAsync(new byte[1], -1, -1, cancelledToken)));
+
+                // buffer is checked first
+                Assert.Throws<ArgumentNullException>("buffer", () =>
+                    FSAssert.CompletesSynchronously(fs.ReadAsync(null, -1, -1, cancelledToken)));
+            }
+        }
+
+        [Fact]
+        public async Task NoopReadAsyncsSucceed()
+        {
+            using (FileStream fs = new FileStream(GetTestFilePath(), FileMode.Create))
+            {
+                // note that these do not succeed synchronously even though they do nothing.
```

Comment:
How common do you think these cases are?  My gut is not very common, but if you disagree, should we fix the product source to special case these to return an already-completed task?